File: Algorithms/Easy/441.Arranging-Coins/ArrangingCoins.go
package easy

// Time complexity: O(logn)
// Space complexity: O(1)
func arrangeCoins(n int) int {
	l, r := 1, n
	for l < r {
		m := (l + r + 1) / 2
		if (m+1)*m/2 <= n {
			l = m
		} else {
			r = m - 1
		}
	}

	return l
}

File: Algorithms/Easy/441.Arranging-Coins/ArrangingCoins.py
# Time complexity: O(logn)
# Space complexity: O(1)
class Solution:
    def arrangeCoins(self, n: int) -> int:
        l, r = 1, n
        while l < r:
            m = (l + r + 1) // 2
            # Sum of m terms of an Arithmetic Progression
            if (m + 1) * m // 2 <= n:
                l = m
            else:
                r = m - 1

        return l
